storagemarket/impl/clientstates/client_fsm.go

```diff
@@ -65,20 +66,39 @@ var ClientEvents = fsm.Events{
 		}),
 	fsm.Event(storagemarket.ClientEventInitiateDataTransfer).
 		From(storagemarket.StorageDealFundsEnsured).To(storagemarket.StorageDealStartDataTransfer),
+
 	fsm.Event(storagemarket.ClientEventUnexpectedDealState).
 		From(storagemarket.StorageDealFundsEnsured).To(storagemarket.StorageDealFailing).
 		Action(func(deal *storagemarket.ClientDeal, status storagemarket.StorageDealStatus, providerMessage string) error {
 			deal.Message = xerrors.Errorf("unexpected deal status while waiting for data request: %d (%s). Provider message: %s", status, storagemarket.DealStates[status], providerMessage).Error()
 			return nil
 		}),
 	fsm.Event(storagemarket.ClientEventDataTransferFailed).
-		FromMany(storagemarket.StorageDealStartDataTransfer, storagemarket.StorageDealTransferring).To(storagemarket.StorageDealFailing).
+		FromMany(storagemarket.StorageDealStartDataTransfer, storagemarket.StorageDealTransferring).
+		To(storagemarket.StorageDealFailing).
 		Action(func(deal *storagemarket.ClientDeal, err error) error {
 			deal.Message = xerrors.Errorf("failed to initiate data transfer: %w", err).Error()
 			return nil
 		}),
+
+	fsm.Event(storagemarket.ClientEventDataTransferRestartFailed).From(storagemarket.StorageDealClientTransferRestart).
+		To(storagemarket.StorageDealFailing).
+		Action(func(deal *storagemarket.ClientDeal, err error) error {
+			deal.Message = xerrors.Errorf("failed to restart data transfer: %w", err).Error()
+			return nil
+		}),
+
 	fsm.Event(storagemarket.ClientEventDataTransferInitiated).
-		From(storagemarket.StorageDealStartDataTransfer).To(storagemarket.StorageDealTransferring),
+		FromMany(storagemarket.StorageDealStartDataTransfer).To(storagemarket.StorageDealTransferring).
+		Action(func(deal *storagemarket.ClientDeal, channelId datatransfer.ChannelID) error {
+			deal.TransferChannelID = channelId
+			return nil
+		}),
+
+	fsm.Event(storagemarket.ClientEventDataTransferRestarted).
+		FromMany(storagemarket.StorageDealClientTransferRestart, storagemarket.StorageDealStartDataTransfer).To(storagemarket.StorageDealTransferring).
```

Comment:
This should not ideally run when in StorageDealStartDataTransfer right? Why is this in the list?

Reply:
@hannahhoward 

We now move from `StorageDealStartDataTransfer ` to `StorageDealTransferring` only after we **see** the `Accept` from the other side.

However, what if the other side sends an accept, we crash before receiving it and then restart. It means the channel exists on both sides and yet, we are restarting in the `StorageDealClientTransferRestart ` state.